NeoML/src/Dnn/DnnSolver.cpp

```diff
@@ -322,57 +297,123 @@ void CDnnSolver::clipGradients(const CObjectArray<CDnnBlob>& paramDiffBlobs)
+		blobSum.LayerOwner = layerTemp;
+	}
+
+	archive >> size;
+	for( int i = 0; i < size; ++i ) {
+		const CString layerPath = convertOldIdToLayerPath( nullptr );
+		SerializeBlobs( mathEngine, archive, layerToGradientHistory.GetOrCreateValue( layerPath ) );
+	}
+}
+
+static const int DnnSolverVersion = 2;
 
-void CDnnSolver::Serialize( CArchive& archive, CDnn& dnn )
+void CDnnSolver::Serialize( CArchive& archive, const CDnn& dnn )
 {
 	const int version = archive.SerializeVersion( DnnSolverVersion );
 	if( archive.IsStoring() ) {
-		CMap<CBaseLayer*, CString> layerPtrToId;
-		mapLayerPtrToId( dnn, layerPtrToId );
-
 		archive << layerToParamDiffBlobsSum.Size();
 		for( int pos = layerToParamDiffBlobsSum.GetFirstPosition(); pos != NotFound;
 			pos = layerToParamDiffBlobsSum.GetNextPosition( pos ) )
 		{
-			archive << layerPtrToId[layerToParamDiffBlobsSum.GetKey( pos )];
+			CString layerPath = layerToParamDiffBlobsSum.GetKey( pos );
+			const CBaseLayer* layer = layerToParamDiffBlobsSum.GetValue( pos ).LayerOwner;
+			NeoAssert( layer != nullptr );
+			CArray<CString> path;
+			layer->GetPath( path );
+			archive.Serialize( path );
+			NeoAssert( path.Size() );
+
 			archive << layerToParamDiffBlobsSum.GetValue( pos ).Count;
 			SerializeBlobs( mathEngine, archive, layerToParamDiffBlobsSum.GetValue( pos ).Sum );
-		}
 
-		archive << layerToGradientHistory.Size();
-		for( int pos = layerToGradientHistory.GetFirstPosition(); pos != NotFound;
-			pos = layerToGradientHistory.GetNextPosition( pos ) )
-		{
-			archive << layerPtrToId[layerToGradientHistory.GetKey( pos )];
-			SerializeBlobs( mathEngine, archive, layerToGradientHistory.GetValue( pos ) );
+			const bool hasGradientHistory = layerToGradientHistory.Has( layerPath );
+			archive << hasGradientHistory;
+			if( hasGradientHistory ) {
+				SerializeBlobs( mathEngine, archive, layerToGradientHistory.GetValue( pos ) );
+			}
 		}
 		archive << learningRate << regularizationL1 << regularizationL2 << maxGradientNorm;
 		archive << clipGradientMin << clipGradientMax;
 	} else {
-		CMap<CString, CBaseLayer*> layerIdToPtr;
-		mapLayerIdToPtr( dnn, layerIdToPtr );
-
 		layerToParamDiffBlobsSum.DeleteAll();
 		layerToGradientHistory.DeleteAll();
 		layersToReduce.DeleteAll();
 		reduceOrder.DeleteAll();
 
-		int size;
-		archive >> size;
-		for( int i = 0; i < size; ++i ) {
-			CString layerId;
-			archive >> layerId;
-			CDiffBlobSum& blobSum = layerToParamDiffBlobsSum.GetOrCreateValue( layerIdToPtr[layerId] );
-			archive >> blobSum.Count;
-			SerializeBlobs( mathEngine, archive, blobSum.Sum );
-		}
-
-		archive >> size;
-		for( int i = 0; i < size; ++i ) {
-			CString layerId;
-			archive >> layerId;
-			SerializeBlobs( mathEngine, archive, layerToGradientHistory.GetOrCreateValue( layerIdToPtr[layerId] ) );
+		if( version >= 2 ) {
+			int size;
+			archive >> size;
+			for( int i = 0; i < size; ++i ) {
+				CArray<CString> path;
+				archive.Serialize( path );
+				NeoAssert( path.Size() );
```

Comment:
```suggestion
				NeoAssert( path.Size() > 0 );
```
явное лучше неявного

Reply:
fixed